flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java

```diff
@@ -551,10 +551,16 @@ private static void sanityCheckShuffleMemory(final Configuration config, final M
 					+ shuffleRangeFraction.maxSize.toString() + "].");
 			}
 			if (isShuffleMemoryFractionExplicitlyConfigured(config) &&
-				!derivedShuffleMemorySize.equals(totalFlinkMemorySize.multiply(shuffleRangeFraction.fraction))) {
-				throw new IllegalConfigurationException("Derived Shuffle Memory size("
-					+ derivedShuffleMemorySize.toString() + ") does not match configured Shuffle Memory fraction ("
-					+ shuffleRangeFraction.fraction + ").");
+				!derivedShuffleMemorySize.equals(totalFlinkMemorySize.multiply(shuffleRangeFraction.fraction)) &&
+				derivedShuffleMemorySize.getBytes() < shuffleRangeFraction.maxSize.getBytes() &&
+				derivedShuffleMemorySize.getBytes() > shuffleRangeFraction.minSize.getBytes()) {
```

Comment:
I can see some corner cases are not covered by this check.

E.g.:
> Total Flink Memory: 100mb
> Shuffle Min: 10mb
> Shuffle Max: 30mb
> Shuffle Fraction: 0.2 or 0.4
> Derived Shuffle Size: 10mb

In the example above, a wrongly derived size that equals one edge of the range (10mb) may pass the sanity check, while the expected size should either be in the middle of the range (20mb when the fraction is 0.2) or capped by the other edge (30mb when the fraction is 0.4).

I think we should also check whether the derived value is capped correctly when it equals to range edge(s). I would suggest something like:
```
if (isShuffleMemoryFractionExplicitlyConfigured(config) {
    final MemorySize shuffleSizeFromFraction = totalFlinkMemorySize.multiply(shuffleRangeFraction.fraction);
    if (shuffleSizeFromFraction.getBytes() <= shuffleRangeFraction.minSize.getBytes()) {
         // check derived size equals to min, or throw an exception
    } else if (shuffleSizeFromFraction.getBytes() >= shuffleRangeFraction.maxSize.getBytes()) {
         // check derived size equals to min, or throw an exception
    } else {
         // check derived size matches the fraction, or throw an exception
    }
}
```

Reply:
Maybe we could take a step back and do it a bit differently. At the moment one could also think of the fraction value as a user wish/soft constraint which does not need to be strictly followed (e.g. if the fraction value is out side of the min max range, then it gets mapped back to it). If this is ok, then one could also argue that if the fraction values lies within the min max range and the actual value does not match the exact fraction value, then this is also ok because otherwise we might not be able to start the cluster. Think of the min max range as the set of valid configurations where the fraction value is the best value. However, if it cannot be fulfilled, then any other value will also be accepted. If this is ok, then the only thing we should do in the sanity check method is to print a logging statement that we could not comply to the fraction value if it differs from the actual shuffle memory.

Reply:
That also works for me. Actually better.

I believe that was the design at the very beginning. Then later we somehow decided that all the explicitly configured values should be respected, or we should fail with a meaningful error message.

I think it would be easier for the users to understand that the fraction is not hard guaranteed, compared to it is respected in some cases (that the derived value falls in the min/max range) while not respected in some other cases (that capped by the min/max constraints).